Get ready for compiling with JDK 1.8 / Java 1.8

googlielmo opened the pull request:
These changes are needed to compile Brooklyn with JDK 1.8, if/when we switch to Java 1.8.
At the same time they work with 1.7, which remains the selected `java.version` for compiling.

 - minor source changes (generics)

aledsage commented on camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java:
```diff
@@ -486,8 +486,8 @@ public DslConfigSupplier(DslComponent component, String keyName) {
 
-            ConfigKey<?> key = targetEntity.getEntityType().getConfigKey(keyName);
-            Maybe<? extends Object> result = targetEntity.config().getNonBlocking(key != null ? key : ConfigKeys.newConfigKey(Object.class, keyName));
+            ConfigKey key = targetEntity.getEntityType().getConfigKey(keyName);
```
Seems a shame to have to strip off the generics. Same below.

googlielmo replied:
@aledsage I agree in principle.
In this particular case, as Java 8 has a deeper / more precise type inference algorithm, it refuses to compile, because it cannot guarantee that the type of the Maybe (the capture of ?) is actually compatible with both the `key` and the `newConfigKey` call. I wonder how it could be permitted in java 7, actually :-)
I will reinstate the generics with Object as that's the only type that's accepted unequivocally.

neykov commented:
I'm able to compile the project successfully without any changes with JDK 1.8.0_45 & 1.8.0_112.

googlielmo commented:
@neykov yes absolutely, the project compiles fine with JDK 8 as long as `java.version` stays at 1.7.
This PR will ensure that whenever we do the switch to Java 1.8, the project will continue to compile without errors.
Will update the description to clarify, thanks!

googlielmo commented:
@aledsage I've addressed your comment and reinstated generics.
In the case of `tryCoerceUntyped` I removed them altogether though. That's because we check types manually on purpose there, and I don't see the point of juggling with wildcard captures and casts just to make the compiler happy! ;-)

neykov commented on camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java:
```diff
@@ -486,8 +486,8 @@ public DslConfigSupplier(DslComponent component, String keyName) {
-            ConfigKey<?> key = targetEntity.getEntityType().getConfigKey(keyName);
-            Maybe<? extends Object> result = targetEntity.config().getNonBlocking(key != null ? key : ConfigKeys.newConfigKey(Object.class, keyName));
+            ConfigKey<Object> key = (ConfigKey<Object>) targetEntity.getEntityType().getConfigKey(keyName);
+            Maybe<Object> result = targetEntity.config().getNonBlocking(key != null ? key : ConfigKeys.newConfigKey(Object.class, keyName));
```
Could instead help the compiler by assigning `key != null ? key : ConfigKeys.newConfigKey(Object.class, keyName)` to an intermediate `ConfigKey<?>` value. Same below.
Given we are casting below anyway could remove generics altogether and stop pretending we are "safe".

neykov commented on utils/common/src/main/java/org/apache/brooklyn/util/javalang/coerce/EnumTypeCoercions.java:
```diff
@@ -70,11 +70,11 @@ public E apply(String input) {
         if (targetType==null) return Maybe.absent("Null enum type");
         if (!targetType.isEnum()) return Maybe.absent("Type '"+targetType+"' is not an enum");
-        return tryCoerce(input, (Class<Enum>)targetType);
+        return tryCoerce(input, targetType);
```
I'd keep the generics as is and add a cast to `(Maybe<T>) ` of the return value (insert before `tryCoerce`).
Credit to Eclipse's quick fix for this one :).
